script/tool/lib/src/publish_check_command.dart

```diff
@@ -121,24 +184,69 @@ class PublishCheckCommand extends PluginCommand {
             'Packages with an SDK constraint on a pre-release of the Dart SDK should themselves be published as a pre-release version.');
   }
 
-  Future<bool> _passesPublishCheck(Directory package) async {
+  Future<_PublishCheckResult> _passesPublishCheck(Directory package) async {
     final String packageName = package.basename;
     print('Checking that $packageName can be published.');
 
     final Pubspec pubspec = _tryParsePubspec(package);
     if (pubspec == null) {
-      return false;
+      print('no pubspec');
+      return _PublishCheckResult.error;
     } else if (pubspec.publishTo == 'none') {
       print('Package $packageName is marked as unpublishable. Skipping.');
-      return true;
+      return _PublishCheckResult._noPublish;
+    }
+
+    final Version version = pubspec.version;
+    final bool alreadyPublished = await _checkIfAlreadyPublished(
+        packageName: packageName, version: version);
+    if (alreadyPublished) {
+      print(
+          'Package $packageName version: $version has already be published on pub.');
+      return _PublishCheckResult._noPublish;
     }
 
     if (await _hasValidPublishCheckRun(package)) {
       print('Package $packageName is able to be published.');
-      return true;
+      return _PublishCheckResult._needsPublish;
     } else {
       print('Unable to publish $packageName');
-      return false;
+      return _PublishCheckResult.error;
     }
   }
+
+  // Check if `packageName` already has `version` published on pub.
+  Future<bool> _checkIfAlreadyPublished(
+      {String packageName, Version version}) async {
+    final PubVersionFinder pubVersionFinder = PubVersionFinder(
+        package: packageName, httpClient: httpClient ?? http.Client());
+    final PubVersionFinderResponse pubVersionFinderResponse =
+        await pubVersionFinder.getPackageVersion();
+    bool published;
+    switch (pubVersionFinderResponse.result) {
+      case PubVersionFinderResult.success:
+        published = pubVersionFinderResponse.versions.contains(version);
+        break;
+      case PubVersionFinderResult.fail:
+        printErrorAndExit(errorMessage: '''
```

Comment:
Won't this break the `--machine` case, because the final message will never print?

Reply:
done.

Reply:
I was thinking this should return something, rather than bailing completely. Is the reason for this that we assume that all the other checks will fail the same way because it's likely a network issue?

If we do want to keep this construction, this line (and the line above) need comments explaining why they are doing what they are doing, since it's non-obvious.

Reply:
ah ok, I updated to print the message and handled the error later